// ionic/dist/commands/ionitron.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_utils_1 = require("@ionic/cli-utils");
const ionitron_1 = require("../lib/ionitron");
let IonitronCommand = class IonitronCommand extends cli_utils_1.Command {
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const locale = options['es'] ? 'es' : 'en';
            const localeStatements = ionitron_1.ionitronStatements[locale];
            const statement = localeStatements[Math.floor(Math.random() * (localeStatements.length))];
            this.env.log.msg(ionitron_1.getIonitronString(statement));
        });
    }
};
IonitronCommand = tslib_1.__decorate([
    cli_utils_1.CommandMetadata({
        name: 'ionitron',
        type: 'global',
        description: 'Print random ionitron messages',
        options: [
            {
                name: 'es',
                description: 'Print in spanish',
                type: Boolean,
            }
        ],
        visible: false,
    })
], IonitronCommand);
exports.IonitronCommand = IonitronCommand;
