// ionic/dist/lib/ssh-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const os = require("os");
const SSHConfig = require("ssh-config");
const cli_utils_1 = require("@ionic/cli-utils");
function loadSSHConfig(p) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const s = yield cli_utils_1.fileToString(p);
        return SSHConfig.parse(s);
    });
}
exports.loadSSHConfig = loadSSHConfig;
function isSSHConfigDirective(entry) {
    return entry && entry.type === SSHConfig.DIRECTIVE;
}
exports.isSSHConfigDirective = isSSHConfigDirective;
function getSSHConfigPath() {
    return path.resolve(os.homedir(), '.ssh', 'config');
}
exports.getSSHConfigPath = getSSHConfigPath;
function findSSHConfigHostSection(conf, host) {
    return conf.find({ Host: host });
}
exports.findSSHConfigHostSection = findSSHConfigHostSection;
function ensureSSHConfigHostAndKeyPath(conf, conn, keyPath) {
    const section = ensureSSHConfigSection(conf, conn.host, conf ? true : false);
    ensureSSHConfigSectionLine(section, 'IdentityFile', keyPath);
    if (typeof conn.port === 'number') {
        ensureSSHConfigSectionLine(section, 'Port', String(conn.port));
    }
}
exports.ensureSSHConfigHostAndKeyPath = ensureSSHConfigHostAndKeyPath;
function ensureSSHConfigSection(conf, host, newline) {
    const section = findSSHConfigHostSection(conf, host);
    if (!section) {
        conf.push(SSHConfig.parse(`${newline ? '\n' : ''}Host ${host}\n`)[0]);
    }
    return conf.find({ Host: host });
}
exports.ensureSSHConfigSection = ensureSSHConfigSection;
function ensureSSHConfigSectionLine(section, key, value) {
    const found = section.config.some((line) => {
        if (isSSHConfigDirective(line)) {
            if (line.param === key) {
                line.value = value;
                return true;
            }
        }
        return false;
    });
    if (!found) {
        section.config = section.config.concat(SSHConfig.parse(`${cli_utils_1.indent(4)}${key} ${value}\n`));
    }
}
exports.ensureSSHConfigSectionLine = ensureSSHConfigSectionLine;
